hw/net/rocker/rocker_p4_rmt.c

```diff
@@ -0,0 +1,216 @@
+                                    char *buf, unsigned short cmd,
+                                    RockerTlv **tlvs);
+
+unsigned char *
+rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
+{
+    // convert iov to a single contiguous buffer for p4 model
+    int len = 0;
+    unsigned char *buf = NULL;
+
+    len = iov_size(iov, iovcnt);
+    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
+    buf = malloc(len);
+    if (buf == NULL) {
+        DPRINTF("Cannot allocate buffer\n");
+        return buf;
+    }
+    iov_to_buf(iov, iovcnt, 0, buf, len);
+    return buf;
+}
+
+
+void rocker_p4_rmt_tx (int eg_port1, void *pkt, int len, int ig_port1, 
+                            void *cookie)
+{
+    struct iovec iov;
+    int    iovcnt = 1;
+    struct rocker *rocker;
+    struct world *w;
+    unsigned int fp_ig_port;
+
+    struct p4_rmt_world *p4_rmt = (struct p4_rmt_world *)cookie; // world_private(w);
+    rocker = p4_rmt->rocker;
+    w = p4_rmt->world;
+
+    fp_port_from_pport((unsigned int)ig_port1, &fp_ig_port);
+
+    DPRINTF("rocker_p4_rmt_tx to port %d, len %d.... \n", eg_port1, len);
+    if (eg_port1 == 0) {
+        DPRINTF("rocker_p4_rmt_tx DROP .... \n");
+        free (pkt);
+        return;
+    }
+    // convert the packet to iov
+    iov.iov_len = len;
+    iov.iov_base = pkt;
+    if (rocker == NULL) {
+        DPRINTF("\"rocker\" not found\n");
+        return;
+    }
+    if (w == NULL) {
+        DPRINTF("P4-RMT world not found\n");
+        return;
+    }
+    // check for sup port
+    if (p4_rmt->is_cpu_port(eg_port1)) {
+        DPRINTF("rocker_p4_rmt_tx to CPU from ingress port %d, len %d.... \n", 
+                    ig_port1, len);
+        rx_produce(w, ig_port1, &iov, iovcnt);
+        return;
+    }
+    // call qemu_send to send out on switch port
+    rocker_port_eg(rocker, eg_port1, &iov, iovcnt);
+    return;
+}
+
+static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt, 
+                            struct desc_info *info,
+                            char *buf, unsigned short cmd,
+                            RockerTlv **tlvs)
+{
+    unsigned int    table_id;
+    unsigned char   *entry;
+    if (tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID] == NULL ||
+        tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY] == NULL) {
+        DPRINTF("Missing TLVs\n");
+        return 0;
+    }
+        
+    table_id = rocker_tlv_get_u32(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID]);
+    entry = rocker_tlv_data(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
+
+    DPRINTF("p4_rmt_table_cmd for table %d\n", table_id);
+
+    if (!p4_rmt->is_table_valid(table_id)) {
+        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
+        return 0;
+    }
+
+    switch (cmd) {
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
+            // check if fn is provided
+            if (p4_rmt->table_ops[table_id].add) {
+                p4_rmt->table_ops[table_id].add(entry);
```

Comment:
Yes it is a blob. One good thing is that, the code to create and consume
the entry  is generated by the p4 tool-chain so bit packing and endian-ness
is taken care of by the generated code. E.g. rocker_p4_l2l3_pd.c which has
entry build functions.

On Sun, Apr 19, 2015 at 10:48 PM, Scott Feldman <notifications@github.com>
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> <https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28665142>:
>
> > +
> > +    table_id = rocker_tlv_get_u32(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID]);
> > +    entry = rocker_tlv_data(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
> > +
> > +    DPRINTF("p4_rmt_table_cmd for table %d\n", table_id);
> > +
> > +    if (!p4_rmt->is_table_valid(table_id)) {
> > +        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
> > +        return 0;
> > +    }
> > +
> > +    switch (cmd) {
> > +        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
> > +            // check if fn is provided
> > +            if (p4_rmt->table_ops[table_id].add) {
> > +                p4_rmt->table_ops[table_id].add(entry);
>
> Hmm...it looks like entry is a binary blob. How is it packed? Are there
> un-aligned access issues? How is it versioned? How is
> big/little/host/network byte order handled? It seems we want TLV-style
> encode/decode of the entry fields to answer all the questions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pbhide/qemu-rocker/pull/1/files#r28665142>.
>